crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Generic.hs

```diff
@@ -1135,29 +1135,43 @@ isAllocatedMut mutOk sym w minAlign (llvmPointerView -> (blk, off)) sz m =
     -- @inThisAllocation a allocatedSz@ produces the predicate that
     -- records whether the pointer @ptr@ of size @sz@ falls within the
     -- allocation of block @a@ of size @allocatedSz@.
-    inThisAllocation :: forall w'. Natural -> Maybe (SymBV sym w') -> IO (Pred sym)
-    inThisAllocation a Nothing = isSameBlock sym blk a
-      -- If the allocation is unbounded, we just have to check we are in the right block
-    inThisAllocation a (Just allocSize)
+    inThisAllocation :: forall w'. Maybe (SymBV sym w') -> IO (Pred sym)
+    inThisAllocation Nothing =
+      case sz of
+        Nothing ->
+          -- Unbounded access of an unbounded allocation must start at offset 0.
+          bvEq sym off =<< bvLit sym w 0
+        Just currSize ->
+          -- Bounded access of an unbounded allocation requires
+          -- that @offset + size <= 2^w@.
+          do zeroSize <- bvEq sym currSize =<< bvLit sym w 0
+             noWrap <- bvUle sym off =<< bvNeg sym currSize
+             orPred sym zeroSize noWrap
+
+    inThisAllocation (Just allocSize)
       -- If the allocation is done at pointer width is equal to @w@, check
       -- if this allocation covers the required range
       | Just Refl <- testEquality w (bvWidth allocSize)
       , Just currSize <- sz =
-        do sameBlock <- isSameBlock sym blk a           -- a == blockname(ptr)
-           smallSize <- bvUle sym currSize allocSize    -- currSize <= allocSize
+        do smallSize <- bvUle sym currSize allocSize    -- currSize <= allocSize
            maxOffset <- bvSub sym allocSize currSize    -- maxOffset = allocSize - currSize
            inRange   <- bvUle sym off maxOffset         -- offset(ptr) <= maxOffset
-           andPred sym sameBlock =<< andPred sym smallSize inRange
-    inThisAllocation a (Just _allocSize)
+           andPred sym smallSize inRange
+
+    inThisAllocation (Just _allocSize)
       -- If the allocation is done at pointer width not equal to @w@,
-      -- check that this allocation is distinct from the base pointer.
-      | otherwise = notPred sym =<< isSameBlock sym blk a
+      -- then this is not the allocation we're looking for. Similarly,
+      -- if @sz@ is @Nothing@ (indicating we are accessing the entire
+      -- address space) then any bounded allocation is too small.
+      | otherwise = return $ falsePred sym
 
     go :: IO (Pred sym) -> [MemAlloc sym] -> IO (Pred sym)
     go fallback [] = fallback
     go fallback (Alloc _ a asz mut alignment _ : r)
       | mutOk mut && alignment >= minAlign =
-        do hereOK <- inThisAllocation a asz
+        do sameBlock <- natEq sym blk =<< natLit sym a
+           inRange <- inThisAllocation asz
+           hereOK <- andPred sym sameBlock inRange
            thereOK <- go fallback r
```

Comment:
We should probably short-cut this call if `hereOK` is concretely true.  Should save time in some common cases.

Reply:
Yeah, I thought about that. Basically what I want is `lazyItePred sameBlock (inThisAllocation asz) (go fallback r)`. Does what4 provide a `lazyItePred` function, or would I have to roll my own?

Reply:
Yeah, there's `iteM` from`What4.Interface` that should do what you want.

Reply:
Ok, thanks, I'll switch it to use that